### ISSUE TITLE
[2018 BMVC Oral] BAM: Bottleneck Attention Module

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#45]
BAM을 사용하면 channel & spatial attention을 얻어 성능을 개선할 수 있다는 논문

논문 내용 중 BAM, experiments만 설명

## 1. Bottleneck Attention Module (BAM)
![image](https://user-images.githubusercontent.com/49643709/147896272-74eedd1f-0323-483a-80e7-aa0c86e288ce.png)
![image](https://user-images.githubusercontent.com/49643709/147896291-0c882df8-0dd1-4174-a480-215b9cb762ed.png)
![image](https://user-images.githubusercontent.com/49643709/147896360-0924239a-eb77-48bf-bede-09186f70b030.png)

- BAM은 input으로 feature map을 받아 output으로 3D attention map을 만들어냄
- BAM은 sigmoid(channel attention + spatial attention)로 구성됨
- Residual learning scheme을 이용하여 original feature map * (1 + BAM(original feature map))을 다음 layer에 전달해줌

### 1.1. Channel attention branch
![image](https://user-images.githubusercontent.com/49643709/147896535-fcb94b2c-4941-4326-a165-fa8770488115.png)

- SENet의 SE block과 매우 유사 (global average pooling + 1 reduction FC + 1 restoration FC)

### 1.2. Spatial attention branch
![image](https://user-images.githubusercontent.com/49643709/147900751-aeb54fcf-e487-43d7-b488-51e81e72db75.png)

- 1*1 conv를 통해 channel reduction (with same reduction ratio r on channel attention branch)
- 2 3*3 dilated conv를 통해 contextual information 얻음 (to aggregate contextual information with a larger receptive field)
- 1*1 conv를 통해 channel reduction (reduced channel to 1)
- BN 추가

### 1.3. Combine 2 attention branches

- Channel, spatial attention map을 기존 input shape로 expand 후 element-wise summation
- Sigmoid function을 사용하여 3D attention map 얻음

### 1.4. Difference between paper and code

[[Official Pytorch Code](https://github.com/Jongchan/attention-module/blob/master/MODELS/bam.py)]

#### 1.4.1. Channel attention branch
![image](https://user-images.githubusercontent.com/49643709/147901440-33ef7b48-48b9-43ff-8ddc-fb4b41d11eda.png)

- 논문 : GAP + FC (reduction) + FC (restoration) + BN
- 코드 : GAP + FC (reduction) + BN + (ReLU) + 1 FC (restoration)
- 즉 BN의 위치가 논문과 다름

#### 1.4.2. Spatial attention branch
![image](https://user-images.githubusercontent.com/49643709/147901668-70296ce9-af8f-4865-83bd-fa89e431c8db.png)

- 논문 : conv (reduction) + conv (dilation) + conv (dilation) + conv (reduction) + BN
- 코드 : conv (reduction) + conv (dilation) + BN + (ReLU) + conv (dilation) + BN + (ReLU) + conv (reduction)
- 즉 BN의 위치가 논문과 다름

#### 1.4.3. Combine 2 attention branches
![image](https://user-images.githubusercontent.com/49643709/147902220-01c26cf2-7907-4a61-b262-ec2821d97ce6.png)

- 논문 : element-wise summation
- 코드 : element-wise product

### 2.1. Ablation studies on CIFAR-100
![image](https://user-images.githubusercontent.com/49643709/147901851-0f14d82c-89e7-4517-bc9b-35e09795aa03.png)

- CIFAR-100에서 찾은 hyperparameter들이 다른 데이터셋에서 최적인지는 알 수 없으나 논문에서는 CIFAR-100에서 구한 값들을 default로 다른 dataset에서 사용

#### 2.1.1. Dilation value and Reduction ratio - Table 1 (a)
- CIFAR-100에서 실험 결과 dilation value 4, reduction ratio 16이 최적임

#### 2.1.2. Separate or Combined branches - Table 1 (b)
- Channel attention branch, spatial attention branch 모두 이용하는 것이 최적임

#### 2.1.3. Combining methods - Table 1 (b)
- Element-wise summation이 최적임
- Element-wise summation : 아무런 정보를 잃지 않고 gradient가 모든 input에 대해 균등하게 배분되기에 효율적임
- Element-wise product : small input에 대해 large gradient를 할당할 수 있어 모델이 converge하기 힘듬
- Element-wise maximum : BAM이 parameter 수가 적은데 gradient가 higher input한테만 가기에 unstable training이 될 수 있음

#### 2.1.4. Comparison with placing original blocks - Table 1 (c)
- BAM을 추가해서 넣은 성능 향상이 depth가 증가해서 생긴 것이 아님을 보여줌

### 2.2. Bottleneck : The efficient point to place BAM 
![image](https://user-images.githubusercontent.com/49643709/147901885-b3fef577-b299-46a4-a431-1d3ba1d5488f.png)

- Conv block 안에 넣기 보다는 bottleneck에 BAM을 넣는 것이 최적임 (PreResNet 제외)

### 2.3. Classification Results & Effectiveness of BAM with Compact Networks
![image](https://user-images.githubusercontent.com/49643709/147901939-ed0a2a1c-8ec1-42a4-93af-269bfea0d739.png)

- CIFAR-100, ImageNet-1K에서 다양한 모델에 넣었을 때 성능 향상이 이루어졌음
- 즉 BAM은 네트워크의 capacity를 효율적으로 늘림 (not due to a naive capacity-increment but effective feature refinement)

### 2.4. MS COCO Object Detection & VOC 2007 Object Detection
![image](https://user-images.githubusercontent.com/49643709/147901952-80167b27-4dbb-4e8d-a8a2-2e3db4a0d671.png)

- BAM만 추가해도 성능 향상이 이루어졌음

### 2.5. Comparison with Squeeze-and-Excitation
![image](https://user-images.githubusercontent.com/49643709/147901967-a96e7568-9e02-48db-87cd-bf3b7e246df6.png)

- SENet 논문에서 제시한 SE block과 비교했을 때, parameter 수가 훨씬 적음에도 불구하고 성능이 더 좋음 (PreResNet 제외)
